### PR TITLE
Documents publishing

### PR DESCRIPTION
**Pull request by voluntadpear**
Adds ability to publish documents and make them read-only. Implements #36 and #37.

Once PR #33 is merged, we should revisit this PR to make the necessary changes to allow users to add comment even on published documents.

### CODE REVIEW

**Comment by voluntadpear** [comment]
Tested that adding comments still works once the document is published. See #33

**Comment by voluntadpear** [comment]
Add new publish route

**Comment by voluntadpear** [comment]
image

This is the current UI for the Publish Confirmation page, any changes to this?

**Comment by mitar** [comment]
Looks great. I would not color cancel red.

Can you change the text to:

> Once it is published it will not be editable anymore. it will become visible to other users and everyone will be able to comment.

File: client/components/publish-confirmation.vue [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      this.$router.push({name: 'document', params: {documentId: this.documentId}});
+    },
+  },
+};
+
+RouterFactory.configure((factory) => {
+  factory.addRoutes([
+    {
+      component,
+      path: '/document/:documentId/publish',
```
Comment (mitar) [inline client/components/publish-confirmation.vue]: Please use `/document/publish/<documentId>`. The reason is that in the future we will have slug appended after the ID.

File: server/content.js [hunk 2]
```diff
@@ -23,10 +24,15 @@ Meteor.methods({
...
-
     let addedCount = 0;
     const latestContent = Content.documents.findOne({contentKey: args.contentKey}, {sort: {version: -1}, fields: {version: 1}});
-
+    const document = Document.documents.findOne({contentKey: args.contentKey});
+    if (document.isPublished()) {
+      // If the document is published we immediately discard this new step
+      // In the future, we should check if the content that they're trying
+      // to add is a comment, to allow that.
+      return 0;
```
Comment (mitar) [inline server/content.js]: But this prevents then commenting, no?

Comment (voluntadpear) [reply server/content.js]: Right now, yes. Comments won't be stored. When PR #33 is merged, we should add comments as an exception, to allow them to be stored.

File: client/components/publish-confirmation.vue (hunk 1 above)
Comment (voluntadpear) [reply client/components/publish-confirmation.vue]: sure, I'll make the change.

**Comment by mitar** [comment]
Please update from master and make commenting work after publishing.

File: client/components/document.vue [hunk 3]
```diff
@@ -92,20 +98,36 @@
...
+        publishDialog: false,
+      };
+    },
     computed: {
       document() {
         return Document.documents.findOne({
           _id: this.documentId,
         });
       },
+      currentUser() {
```
Comment (mitar) [inline client/components/document.vue]: I think we could extract this into a plugin. And then have `this.currentUser` returning `Meteor.user({username: 1, avatar: 1})`, and `this.currentUserId` returning `Meteor.userId()`. The latter is then useful when you just want to check if user is logged in.

File: server/content.js (hunk 2 above)
Comment (mitar) [reply server/content.js]: Return `addedCount` here (which is 0).

File: client/components/publish-confirmation.vue [hunk 4]
```diff
@@ -0,0 +1,69 @@
+<template>
```
Comment (mitar) [inline client/components/publish-confirmation.vue]: Probably the whole template should render something like 404 or 403 if user is not logged in (and in the future if no permissions).

File: client/components/publish-confirmation.vue [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      this.$router.push({name: 'document', params: {documentId: this.documentId}});
+    },
+    onPublishClick() {
+      if (!this.currentUser) {
+        // only publish article if current user is set
+        this.$router.push({name: 'document', params: {documentId: this.documentId}});
+        return;
+      }
+      Document.publish({documentId: this.documentId});
+      this.$router.push({name: 'document', params: {documentId: this.documentId}});
```
Comment (mitar) [inline client/components/publish-confirmation.vue]: Post a snackbar message if document has been published successfully, or if not. (Use callback from publish call.)

File: client/components/publish-confirmation.vue [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+          <v-card>
+            <v-card-title primary-title>
+              <div>
+                <h3>Do you wish to publish this document now?</h3>
+                <div>Once it's published it will not be editable anymore.</div>
+              </div>
+            </v-card-title>
+            <v-card-actions>
+              <v-btn flat color="primary" class="mx-0" @click="onCancelClick">Cancel</v-btn>
+              <v-btn flat color="primary" class="mx-0" @click="onPublishClick">Publish</v-btn>
```
Comment (mitar) [inline client/components/publish-confirmation.vue]: Disable the form while Meteor method call is being processed. See `user-signin` for an example.

Moreover, this call should probably happen only on the server, not on the client. So no simulation here. User does not expect latency compensation.

File: client/components/document.vue (hunk 3 above)
Comment (voluntadpear) [reply client/components/document.vue]: That's a good idea. Just to be clear, you are talking about a Vue plugin that we initialize doing something like `Vue.use(MeteorUserPlugin)`?

File: client/components/publish-confirmation.vue (hunk 4 above)
Comment (voluntadpear) [reply client/components/publish-confirmation.vue]: We could also add a route interceptor and redirect to the `NotFound` component or a new one for unauthorized access.

Comment (mitar) [reply client/components/publish-confirmation.vue]: See the user-signing component. I have this there.

Comment (mitar) [reply client/components/publish-confirmation.vue]: I would not for now do it at a route level, just we render access denied if user is not logged in. The opposite of the user-signing component.

File: client/components/document.vue (hunk 3 above)
Comment (mitar) [reply client/components/document.vue]: Yea.

File: client/components/publish-confirmation.vue (hunk 4 above)
Comment (voluntadpear) [reply client/components/publish-confirmation.vue]: Got it!

**Comment by voluntadpear** [comment]
This one is ready for review

**Pull request merged**